### PR TITLE
Remove the `TypedArena::alloc_from_iter` specialization.

### PR DESCRIPTION
**Pull request by nnethercote**
It was added in #78569. It's complicated and doesn't actually help
performance.

r? @cjgillot

### CODE REVIEW

**Comment by nnethercote** [comment]
Perf is unchanged, I think this is good to go.

r? @cjgillot

File: compiler/rustc_arena/src/lib.rs [hunk 1]
```diff
@@ -270,8 +198,35 @@ impl<T> TypedArena<T> {
...
+        // iterator one at a time into the allocated memory. That's easy
+        // because the elements don't implement `Drop`. But for `TypedArena`
+        // they do implement `Drop`, which means that if the iterator panics we
+        // could end up with some allocated-but-uninitialized elements, which
+        // will then cause UB in `TypedArena::drop`.
+        //
+        // Instead we use an approach where any iterator panic will occur
+        // before the memory is allocated. This function is much less hot than
+        // `DroplessArena::alloc_from_iter`, so it doesn't need to be
+        // hyper-optimized.
```
Comment (cjgillot) [inline compiler/rustc_arena/src/lib.rs:215]: Do we need a fast path if `!needs_drop::<T>()`? That would call the drop-less codepath and eventually benefit from the opt.

Comment (nnethercote) [reply compiler/rustc_arena/src/lib.rs:215]: Not sure I understand the question. Any `T` not implementing `Drop` will be allocated in the `DroplessArena`, as per [this comment](https://github.com/rust-lang/rust/blob/eea26141ec2bf07a825011af841463dec51a04c9/compiler/rustc_arena/src/lib.rs#L603-L615).

**Pull request merged**